src/cpu/core_dyn_x86.cpp

```diff
@@ -248,13 +257,34 @@ static void dyn_restoreregister(DynReg * src_reg, DynReg * dst_reg) {
 
 #include "core_dyn_x86/decoder.h"
 
-static void copy_dh_fpu_to_normal()
+#	if defined(X86_DYNFPU_DH_ENABLED)
+
+static void maybe_copy_host_fpu_to_dh()
 {
-	FPU_SetTag(dyn_dh_fpu.state.tag);
-	FPU_SetCW(dyn_dh_fpu.state.cw);
-	FPU_SetSW(dyn_dh_fpu.state.sw);
+	if (dyn_dh_fpu.state_used) {
+		gen_dh_fpu_save();
+		dyn_dh_fpu.state_copied_to_normal = false;
+	}
+}
+
+static void maybe_copy_dh_fpu_to_normal()
+{
+	if (dyn_dh_fpu.state_copied_to_normal) {
+		// LOG_MSG("FPU: skipping dh to normal because dh hasn't changed (normal FPU still valid)");
+		return;
+	}
+
+	// Pre-requisite: host FPU should be copied to the dh FPU
+	assert(!dyn_dh_fpu.state_used);
+
+	FPU_SetTag(static_cast<uint16_t>(dyn_dh_fpu.state.tag & 0xffff));
+	FPU_SetCW(static_cast<uint16_t>(dyn_dh_fpu.state.cw & 0xffff));
+	FPU_SetSW(static_cast<uint16_t>(dyn_dh_fpu.state.sw & 0xffff));
 	TOP = FPU_GET_TOP();
-	FPU_LoadPRegs(&dyn_dh_fpu.state.st_reg[0][0]);
+	FPU_SetPRegsFrom(dyn_dh_fpu.state.st_reg);
+
+	dyn_dh_fpu.state_copied_to_normal = true;
+	// LOG_MSG("FPU: copying dh to normal because dh has changed (normal FPU now valid)");
 }
 
 static void copy_normal_fpu_to_dh()
```

Comment:
It's not safe to unconditionally copy normal fpu back.

Reply:
We can only copy normal fpu back when we previously copied it in.  Otherwise we risk overwriting dynamic fpu with unknown state.